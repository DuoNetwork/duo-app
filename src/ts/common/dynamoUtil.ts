import { DynamoUtil } from '@finbook/duo-market-data';

let config = {};
try {
	config = require(`../keys/aws.ui.${__KOVAN__ ? 'dev' : 'live'}.json`);
} catch (error) {
	console.log(error);
}

const dynamoUtil = new DynamoUtil(config, !__KOVAN__, '');

export default dynamoUtil;
