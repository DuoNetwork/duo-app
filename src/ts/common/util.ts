import { IContractPrice } from '@finbook/duo-contract-wrapper';
import { IPriceStatus, IStatus } from '@finbook/duo-market-data';
import * as d3 from 'd3';
import moment, { DurationInputArg2 } from 'moment';
import { ISourceData } from './types';

class Util {
	public convertUpdateTime(timestamp: number): string {
		const diff = this.getUTCNowTimestamp() - timestamp;
		if (diff < 60000) return 'Just Now';
		else if (diff < 3600000) return Math.floor(diff / 60000) + ' Minutes Ago';
		else if (diff < 86400000) return Math.floor(diff / 3600000) + ' Hours Ago';
		else if (diff < 2592000000) return Math.floor(diff / 86400000) + ' Days Ago';
		else return 'Long Time Ago';
	}

	public formatTime(e: number) {
		return moment(e).format('YYYY-MM-DD HH:mm:ss');
	}

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public getDates(length: number, step: number, stepSize: DurationInputArg2, format: string) {
		const dates: string[] = [];
		const date = moment.utc();
		for (let i = 0; i < length; i++) {
			dates.push(date.format(format));
			date.subtract(step, stepSize);
		}
		dates.sort((a, b) => a.localeCompare(b));

		return dates;
	}

	public round(num: number) {
		return +(Math.floor((num + 'e+8') as any) + 'e-8');
	}

	public formatBalance(num: number) {
		if (Math.abs(num) < 1e-8) return '0.000';
		return d3
			.format(Math.abs(num) > 1 ? ',.4s' : ',.4n')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatNumber(num: number) {
		if (Math.abs(num) < 1e-8) return '0.000';
		if (Math.abs(num) < 1) return d3.format(',.4n')(num);
		if (Math.abs(num) < 100000) return d3.format(',.2f')(num);
		return d3
			.format(',.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public getLastPriceFromStatus(status: IStatus[]): ISourceData<IContractPrice> {
		// const bitfinex: ICustodianPrice = {
		// 	price: 0,
		// 	timestamp: 0
		// };
		const kraken: IContractPrice = {
			price: 0,
			timestamp: 0
		};
		const gemini: IContractPrice = {
			price: 0,
			timestamp: 0
		};
		const gdax: IContractPrice = {
			price: 0,
			timestamp: 0
		};
		status.forEach(s => {
			// if (s.process === 'PRICE_AWS_PUBLIC_BITFINEX') {
			// 	bitfinex.price = (s as IPriceStatus).price;
			// 	bitfinex.timestamp = (s as IPriceStatus).timestamp;
			// } else
			if (s.process === 'TRADE_AWS_PUBLIC_GEMINI') {
				gemini.price = (s as IPriceStatus).price;
				gemini.timestamp = (s as IPriceStatus).timestamp;
			} else if (s.process === 'TRADE_AWS_PUBLIC_KRAKEN') {
				kraken.price = (s as IPriceStatus).price;
				kraken.timestamp = (s as IPriceStatus).timestamp;
			} else if (s.process === 'TRADE_AWS_PUBLIC_GDAX') {
				gdax.price = (s as IPriceStatus).price;
				gdax.timestamp = (s as IPriceStatus).timestamp;
			}
		});

		return {
			// bitfinex,
			kraken,
			gemini,
			gdax
		};
	}
}

const util = new Util();
export default util;
