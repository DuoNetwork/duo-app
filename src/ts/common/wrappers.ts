import {
	Constants as WrapperConstants,
	DualClassWrapper,
	EsplanadeWrapper,
	ICustodianWrappers,
	MagiWrapper,
	StakeWrapper,
	Web3Wrapper
} from '@finbook/duo-contract-wrapper';

export const web3Wrapper = new Web3Wrapper(
	window,
	(__KOVAN__ ? WrapperConstants.PROVIDER_INFURA_KOVAN : WrapperConstants.PROVIDER_INFURA_MAIN).replace('/v3', ''),
	'',
	!__KOVAN__
);
export const dualClassWrappers: ICustodianWrappers = {};
for (const type in web3Wrapper.contractAddresses.Custodians) {
	dualClassWrappers[type] = {};
	for (const tenor in web3Wrapper.contractAddresses.Custodians[type])
		dualClassWrappers[type][tenor] = new DualClassWrapper(
			web3Wrapper,
			web3Wrapper.contractAddresses.Custodians[type][tenor].custodian.address
		);
}

export const getDualClassWrapperByTypeTenor = (type: string, tenor: string) => {
	if (dualClassWrappers[type] && dualClassWrappers[type][tenor])
		return dualClassWrappers[type][tenor];
	return dualClassWrappers[WrapperConstants.BEETHOVEN][WrapperConstants.TENOR_PPT];
};

export const getDualClassAddressByTypeTenor = (type: string, tenor: string) => {
	if (web3Wrapper.contractAddresses.Custodians[type][tenor])
		return web3Wrapper.contractAddresses.Custodians[type][tenor];
	return web3Wrapper.contractAddresses.Custodians[WrapperConstants.BEETHOVEN][
		WrapperConstants.TENOR_PPT
	];
};

export const magiWrapper = new MagiWrapper(
	web3Wrapper,
	web3Wrapper.contractAddresses.Oracles[0].address
);

export const esplanadeWrapper = new EsplanadeWrapper(
	web3Wrapper,
	web3Wrapper.contractAddresses.MultiSigManagers[0].address
);

export const stakeWrapper0 = new StakeWrapper(
	web3Wrapper,
	web3Wrapper.contractAddresses.Stakes[0].address
);

export const stakeWrapper1 = new StakeWrapper(
	web3Wrapper,
	web3Wrapper.contractAddresses.Stakes[1].address
);

export const stakeWrappers = [stakeWrapper0, stakeWrapper1];

export const calculateNav = DualClassWrapper.calculateNav;

export const getTokenInterestOrLeverage = DualClassWrapper.getTokenInterestOrLeverage;
