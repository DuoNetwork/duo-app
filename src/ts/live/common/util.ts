import * as d3 from 'd3';
import moment, { DurationInputArg2 } from 'moment';
import * as CST from './constants';
import { ICustodianPrice, IPriceStatus, ISourceData, IStatus } from './types';

class Util {
	public convertUpdateTime(timestamp: number): string {
		const diff = this.getUTCNowTimestamp() - timestamp;
		if (diff < 60000) return 'Just Now';
		else if (diff < 3600000) return Math.floor(diff / 60000) + ' Minutes Ago';
		else if (diff < 86400000) return Math.floor(diff / 3600000) + ' Hours Ago';
		else if (diff < 2592000000) return Math.floor(diff / 86400000) + ' Days Ago';
		else return 'Long Time Ago';
	}

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public calculateNav(
		price: number,
		time: number,
		resetPrice: number,
		resetTime: number,
		alpha: number,
		beta: number,
		period: number,
		coupon: number
	) {
		const navParent = (price / resetPrice / beta) * (1 + alpha);

		const navA = 1 + Math.floor((time - resetTime) / 1000 / period) * coupon;
		const navAAdj = navA * alpha;
		if (navParent <= navAAdj) return [navParent / alpha, 0];
		else return [navA, navParent - navAAdj];
	}

	public getDates(length: number, step: number, stepSize: DurationInputArg2, format: string) {
		const dates: string[] = [];
		const date = moment.utc();
		for (let i = 0; i < length; i++) {
			dates.push(date.format(format));
			date.subtract(step, stepSize);
		}
		dates.sort((a, b) => a.localeCompare(b));

		return dates;
	}

	public round(num: number) {
		return +(Math.floor((num + 'e+8') as any) + 'e-8');
	}

	public range(start: number, end: number) {
		const subArray = Array.apply(null, { length: end }).map(Number.call, Number);
		return subArray.slice(start, end);
	}

	public formatBalance(num: number) {
		if (Math.abs(num) < 1e-8) return '0.000';
		return d3
			.format(Math.abs(num) > 1 ? ',.4s' : ',.4n')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatNumber(num: number) {
		if (Math.abs(num) < 1e-8) return '0.000';
		if (Math.abs(num) < 1) return d3.format(',.4n')(num);
		if (Math.abs(num) < 100000) return d3.format(',.2f')(num);
		return d3
			.format(',.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public getCurrentPath(location: any) {
		return location.pathname === '/' ? location.pathname + 'beethoven' : location.pathname;
	}

	public getLastPriceFromStatus(status: IStatus[]): ISourceData<ICustodianPrice> {
		// const bitfinex: ICustodianPrice = {
		// 	address: CST.DUMMY_ADDR,
		// 	price: 0,
		// 	timestamp: 0
		// };
		const kraken: ICustodianPrice = {
			address: CST.DUMMY_ADDR,
			price: 0,
			timestamp: 0
		};
		const gemini: ICustodianPrice = {
			address: CST.DUMMY_ADDR,
			price: 0,
			timestamp: 0
		};
		const gdax: ICustodianPrice = {
			address: CST.DUMMY_ADDR,
			price: 0,
			timestamp: 0
		};
		status.forEach(s => {
			// if (s.process === 'PRICE_AWS_PUBLIC_BITFINEX') {
			// 	bitfinex.price = (s as IPriceStatus).price;
			// 	bitfinex.timestamp = (s as IPriceStatus).timestamp;
			// } else
			if (s.process === 'TRADE_AWS_PUBLIC_GEMINI') {
				gemini.price = (s as IPriceStatus).price;
				gemini.timestamp = (s as IPriceStatus).timestamp;
			} else if (s.process === 'TRADE_AWS_PUBLIC_KRAKEN') {
				kraken.price = (s as IPriceStatus).price;
				kraken.timestamp = (s as IPriceStatus).timestamp;
			} else if (s.process === 'TRADE_AWS_PUBLIC_GDAX') {
				gdax.price = (s as IPriceStatus).price;
				gdax.timestamp = (s as IPriceStatus).timestamp;
			}
		});

		return {
			// bitfinex,
			kraken,
			gemini,
			gdax
		};
	}
}

const util = new Util();
export default util;
