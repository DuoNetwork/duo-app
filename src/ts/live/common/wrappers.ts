import BeethovenWapper from '../../../../../duo-contract-wrapper/src/BeethovenWapper';
import Web3Wrapper from '../../../../../duo-contract-wrapper/src/Web3Wrapper';
import infura from '../../keys/infura.json';
import * as CST from './constants';

const provider =
	(__KOVAN__ ? CST.PROVIDER_INFURA_KOVAN : CST.PROVIDER_INFURA_MAIN) + '/' + infura.token;
export const web3Wrapper = new Web3Wrapper(window, '', provider, !__KOVAN__);
export const beethovenWapper = new BeethovenWapper(web3Wrapper);
