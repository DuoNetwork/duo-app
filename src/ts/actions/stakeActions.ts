import { IStakeAddress, IStakeLot, IStakeStates } from '@finbook/duo-contract-wrapper';
//import moment from 'moment';
import * as CST from 'ts/common/constants';
import { VoidThunkAction } from 'ts/common/types';
//import util from 'ts/common/util';
import { stakeWrapper, web3Wrapper } from 'ts/common/wrappers';

export function statesUpdate(states: IStakeStates) {
	return {
		type: CST.AC_STK_STATES,
		value: states
	};
}

export function getStates(): VoidThunkAction {
	return async dispatch => {
		const states = await stakeWrapper.getStates();
		dispatch(statesUpdate(states));
	};
}

export function balancesUpdate(duo: number) {
	return {
		type: CST.AC_STK_BALANCE,
		value: duo
	};
}

export function getBalances(): VoidThunkAction {
	return async (dispatch, getState) => {
		const account = getState().web3.account;
		const duoBalance = await web3Wrapper.getErc20Balance(
			web3Wrapper.contractAddresses.DUO,
			account
		);
		dispatch(balancesUpdate(duoBalance));
	};
}

export function addressesUpdate(addr: IStakeAddress) {
	return {
		type: CST.AC_STK_ADDRESSES,
		value: addr
	};
}

export function getAddresses(): VoidThunkAction {
	return async dispatch => {
		const addr = await stakeWrapper.getAddresses();
		dispatch(addressesUpdate(addr));
	};
}

export function userStakeUpdate(userStake: { [key: string]: IStakeLot[] }) {
	return {
		type: CST.AC_STK_USERSTAKE,
		value: userStake
	};
}

export function getUserStake(): VoidThunkAction {
	return async (dispatch, getState) => {
		const account = getState().web3.account;
		const pfList = await stakeWrapper.getPfList();
		const userStake = await stakeWrapper.getUserStakes(account, pfList);
		dispatch(userStakeUpdate(userStake));
	};
}

export function oracleStakeUpdate(oracleStake: { [key: string]: number }) {
	return {
		type: CST.AC_STK_ORACLESTAKE,
		value: oracleStake
	};
}

export function getOracleStake(): VoidThunkAction {
	return async dispatch => {
		const oracleStake = await stakeWrapper.getOracleStakes();
		dispatch(oracleStakeUpdate(oracleStake));
	};
}

export function userAwardUpdate(userAward: number) {
	return {
		type: CST.AC_STK_AWARD,
		value: userAward
	};
}

export function getUserAward(): VoidThunkAction {
	return async (dispatch, getState) => {
		console.log('Getting Award');
		const account = getState().web3.account;
		const userAward = await stakeWrapper.getUserAward(account);
		console.log('User award: ' + userAward);
		dispatch(userAwardUpdate(userAward));
	};
}

export function subscriptionUpdate(intervalId: number) {
	return {
		type: CST.AC_STK_SUB,
		id: intervalId
	};
}

export function refresh(): VoidThunkAction {
	return async dispatch => {
		await dispatch(getStates());
		dispatch(getBalances());
		dispatch(getAddresses());
		dispatch(getUserStake());
		dispatch(getOracleStake());
		dispatch(getUserAward());
	};
}

export function subscribe(): VoidThunkAction {
	return async dispatch => {
		dispatch(subscriptionUpdate(0));
		dispatch(refresh());
		dispatch(subscriptionUpdate(window.setInterval(() => dispatch(refresh()), 30000)));
	};
}

export function refreshAdmin(): VoidThunkAction {
	return async dispatch => {
		await dispatch(getStates());
		dispatch(getAddresses());
	};
}

export function subscribeAdmin(): VoidThunkAction {
	return async dispatch => {
		dispatch(subscriptionUpdate(0));
		dispatch(refreshAdmin());
		dispatch(subscriptionUpdate(window.setInterval(() => dispatch(refreshAdmin()), 30000)));
	};
}
