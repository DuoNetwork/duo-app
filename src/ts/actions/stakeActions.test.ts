// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import util from 'ts/common/util';
import { stakeWrapper } from 'ts/common/wrappers';
import * as stakeActions from './stakeActions';

const stakeStore = configureMockStore([thunk]);

describe('actions', () => {
	stakeWrapper.getStates = jest.fn(() =>
		Promise.resolve({
			canStake: false,
			canUnstake: false,
			lockMinTimeInSecond: 0,
			minStakeAmt: 0,
			maxStakePerOracle: 0,
			totalAwardsToDistribute: 0
		})
	);
	stakeWrapper.getAddresses = jest.fn(() =>
		Promise.resolve({
			operator: '0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0',
			priceFeedList: [
				'0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a',
				'0x002002812b42601Ae5026344F0395E68527bb0F8',
				'0x00476E55e02673B0E4D2B474071014D5a366Ed4E'
			]
		})
	);

	test('balancesUpdate', () => {
		expect(stakeActions.balancesUpdate(123)).toMatchSnapshot();
	});

	test('getBalances', () => {
		const store: any = stakeStore({
			web3: { account: '0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0' }
		});
		store.dispatch(stakeActions.getBalances());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});
	test('addressesUpdate', () => {
		expect(
			stakeActions.addressesUpdate({
				operator: '0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0',
				priceFeedList: [
					'0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a',
					'0x002002812b42601Ae5026344F0395E68527bb0F8',
					'0x00476E55e02673B0E4D2B474071014D5a366Ed4E'
				]
			})
		).toMatchSnapshot();
	});
	test('getAddresses', () => {
		const store: any = stakeStore({
			web3: { account: '0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0' }
		});
		store.dispatch(stakeActions.getAddresses());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});
	test('userStakeUpdate', () => {
		expect(stakeActions.userStakeUpdate({} as any)).toMatchSnapshot();
	});

	test('getUserStake', () => {
		const store: any = stakeStore({
			web3: { account: '0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0' }
		});
		store.dispatch(stakeActions.getUserStake());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});
	test('userAwardUpdate', () => {
		expect(stakeActions.userAwardUpdate({} as any)).toMatchSnapshot();
	});

	test('getUserAward', () => {
		const store: any = stakeStore({
			web3: { account: '0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0' }
		});
		store.dispatch(stakeActions.getUserAward());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});
	test('oracleStakeUpdate', () => {
		expect(stakeActions.oracleStakeUpdate({} as any)).toMatchSnapshot();
	});

	test('getOracleStake', () => {
		const store: any = stakeStore({
			web3: { account: '0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0' }
		});
		store.dispatch(stakeActions.getOracleStake());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});
	test('contractDUOUpdate', () => {
		expect(stakeActions.contractDUOUpdate({} as any)).toMatchSnapshot();
	});

	test('gerContractDUO', () => {
		const store: any = stakeStore({
			web3: { account: '0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0' }
		});
		store.dispatch(stakeActions.gerContractDUO());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});
	test('refresh', () => {
		util.getDates = jest.fn(() => ['1970-01-15']);
		const store: any = stakeStore({
			stake: {
				states: {},
				addresses: []
			},
			web3: { account: '0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0' }
		});

		store.dispatch(stakeActions.refresh());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('subscriptionUpdate', () => {
		expect(stakeActions.subscriptionUpdate(123)).toMatchSnapshot();
	});

	test('subscribe', () => {
		window.setInterval = jest.fn(() => 123);
		const store: any = stakeStore({
			web3: { account: '0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0' }
		});
		store.dispatch(stakeActions.subscribe());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				expect((window.setInterval as jest.Mock).mock.calls[0][1]).toMatchSnapshot();

				(window.setInterval as jest.Mock).mock.calls[0][0]();
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});
});
