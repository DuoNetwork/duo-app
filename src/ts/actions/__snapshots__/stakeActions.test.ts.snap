// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions addressesUpdate 1`] = `
Object {
  "index": 0,
  "type": "stakeAddresses",
  "value": Object {
    "operator": "0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0",
    "priceFeedList": Array [
      "0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a",
      "0x002002812b42601Ae5026344F0395E68527bb0F8",
      "0x00476E55e02673B0E4D2B474071014D5a366Ed4E",
    ],
  },
}
`;

exports[`actions allowanceUpdate 1`] = `
Object {
  "index": 0,
  "type": "stakeAllowance",
  "value": 123,
}
`;

exports[`actions balancesUpdate 1`] = `
Object {
  "type": "stakeBalance",
  "value": 123,
}
`;

exports[`actions contractDUOUpdate 1`] = `
Object {
  "index": 0,
  "type": "contractDUO",
  "value": Object {},
}
`;

exports[`actions gerContractDUO 1`] = `
Array [
  Object {
    "index": 0,
    "type": "contractDUO",
    "value": 100,
  },
]
`;

exports[`actions getAddresses 1`] = `
Array [
  Object {
    "index": 0,
    "type": "stakeAddresses",
    "value": Object {
      "operator": "0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0",
      "priceFeedList": Array [
        "0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a",
        "0x002002812b42601Ae5026344F0395E68527bb0F8",
        "0x00476E55e02673B0E4D2B474071014D5a366Ed4E",
      ],
    },
  },
]
`;

exports[`actions getAllowance 1`] = `
Array [
  Object {
    "index": 0,
    "type": "stakeAllowance",
    "value": 100,
  },
]
`;

exports[`actions getBalances 1`] = `
Array [
  Object {
    "type": "stakeBalance",
    "value": 100,
  },
]
`;

exports[`actions getOracleStake 1`] = `
Array [
  Object {
    "index": 0,
    "type": "orcaleStake",
    "value": Object {
      "0x002002812b42601Ae5026344F0395E68527bb0F8": 234,
      "0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a": 123,
    },
  },
]
`;

exports[`actions getUserAward 1`] = `
Array [
  Object {
    "index": 0,
    "type": "userAward",
    "value": 500,
  },
]
`;

exports[`actions getUserStake 1`] = `
Array [
  Object {
    "index": 0,
    "type": "userStake",
    "value": Object {
      "0x002002812b42601Ae5026344F0395E68527bb0F8": Array [
        Object {
          "amount": 123,
          "timestamp": 1234567890,
        },
      ],
      "0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a": Array [
        Object {
          "amount": 123,
          "timestamp": 1234567890,
        },
      ],
    },
  },
]
`;

exports[`actions oracleStakeUpdate 1`] = `
Object {
  "index": 0,
  "type": "orcaleStake",
  "value": Object {},
}
`;

exports[`actions refresh 1`] = `
Array [
  Object {
    "index": 0,
    "type": "stakeStates",
    "value": Object {
      "canStake": false,
      "canUnstake": false,
      "lockMinTimeInSecond": 0,
      "maxStakePerOracle": 0,
      "minStakeAmt": 0,
      "totalAwardsToDistribute": 0,
    },
  },
  Object {
    "type": "stakeBalance",
    "value": 100,
  },
  Object {
    "index": 0,
    "type": "stakeAllowance",
    "value": 100,
  },
  Object {
    "index": 0,
    "type": "stakeAddresses",
    "value": Object {
      "operator": "0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0",
      "priceFeedList": Array [
        "0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a",
        "0x002002812b42601Ae5026344F0395E68527bb0F8",
        "0x00476E55e02673B0E4D2B474071014D5a366Ed4E",
      ],
    },
  },
  Object {
    "index": 0,
    "type": "userAward",
    "value": 500,
  },
  Object {
    "index": 0,
    "type": "contractDUO",
    "value": 100,
  },
  Object {
    "index": 0,
    "type": "userStake",
    "value": Object {
      "0x002002812b42601Ae5026344F0395E68527bb0F8": Array [
        Object {
          "amount": 123,
          "timestamp": 1234567890,
        },
      ],
      "0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a": Array [
        Object {
          "amount": 123,
          "timestamp": 1234567890,
        },
      ],
    },
  },
  Object {
    "index": 0,
    "type": "orcaleStake",
    "value": Object {
      "0x002002812b42601Ae5026344F0395E68527bb0F8": 234,
      "0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a": 123,
    },
  },
]
`;

exports[`actions refreshAdmin 1`] = `
Array [
  Object {
    "index": 0,
    "type": "stakeStates",
    "value": Object {
      "canStake": false,
      "canUnstake": false,
      "lockMinTimeInSecond": 0,
      "maxStakePerOracle": 0,
      "minStakeAmt": 0,
      "totalAwardsToDistribute": 0,
    },
  },
  Object {
    "index": 0,
    "type": "stakeAddresses",
    "value": Object {
      "operator": "0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0",
      "priceFeedList": Array [
        "0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a",
        "0x002002812b42601Ae5026344F0395E68527bb0F8",
        "0x00476E55e02673B0E4D2B474071014D5a366Ed4E",
      ],
    },
  },
]
`;

exports[`actions subscribe 1`] = `
Array [
  Object {
    "id": 0,
    "type": "stakeSubscription",
  },
  Object {
    "id": 123,
    "type": "stakeSubscription",
  },
  Object {
    "index": 0,
    "type": "stakeStates",
    "value": Object {
      "canStake": false,
      "canUnstake": false,
      "lockMinTimeInSecond": 0,
      "maxStakePerOracle": 0,
      "minStakeAmt": 0,
      "totalAwardsToDistribute": 0,
    },
  },
  Object {
    "type": "stakeBalance",
    "value": 100,
  },
  Object {
    "index": 0,
    "type": "stakeAllowance",
    "value": 100,
  },
  Object {
    "index": 0,
    "type": "stakeAddresses",
    "value": Object {
      "operator": "0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0",
      "priceFeedList": Array [
        "0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a",
        "0x002002812b42601Ae5026344F0395E68527bb0F8",
        "0x00476E55e02673B0E4D2B474071014D5a366Ed4E",
      ],
    },
  },
  Object {
    "index": 0,
    "type": "userAward",
    "value": 500,
  },
  Object {
    "index": 0,
    "type": "contractDUO",
    "value": 100,
  },
  Object {
    "index": 0,
    "type": "userStake",
    "value": Object {
      "0x002002812b42601Ae5026344F0395E68527bb0F8": Array [
        Object {
          "amount": 123,
          "timestamp": 1234567890,
        },
      ],
      "0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a": Array [
        Object {
          "amount": 123,
          "timestamp": 1234567890,
        },
      ],
    },
  },
  Object {
    "index": 0,
    "type": "orcaleStake",
    "value": Object {
      "0x002002812b42601Ae5026344F0395E68527bb0F8": 234,
      "0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a": 123,
    },
  },
]
`;

exports[`actions subscribe 2`] = `30000`;

exports[`actions subscribe 3`] = `
Array [
  Object {
    "id": 0,
    "type": "stakeSubscription",
  },
  Object {
    "id": 123,
    "type": "stakeSubscription",
  },
  Object {
    "index": 0,
    "type": "stakeStates",
    "value": Object {
      "canStake": false,
      "canUnstake": false,
      "lockMinTimeInSecond": 0,
      "maxStakePerOracle": 0,
      "minStakeAmt": 0,
      "totalAwardsToDistribute": 0,
    },
  },
  Object {
    "type": "stakeBalance",
    "value": 100,
  },
  Object {
    "index": 0,
    "type": "stakeAllowance",
    "value": 100,
  },
  Object {
    "index": 0,
    "type": "stakeAddresses",
    "value": Object {
      "operator": "0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0",
      "priceFeedList": Array [
        "0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a",
        "0x002002812b42601Ae5026344F0395E68527bb0F8",
        "0x00476E55e02673B0E4D2B474071014D5a366Ed4E",
      ],
    },
  },
  Object {
    "index": 0,
    "type": "userAward",
    "value": 500,
  },
  Object {
    "index": 0,
    "type": "contractDUO",
    "value": 100,
  },
  Object {
    "index": 0,
    "type": "userStake",
    "value": Object {
      "0x002002812b42601Ae5026344F0395E68527bb0F8": Array [
        Object {
          "amount": 123,
          "timestamp": 1234567890,
        },
      ],
      "0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a": Array [
        Object {
          "amount": 123,
          "timestamp": 1234567890,
        },
      ],
    },
  },
  Object {
    "index": 0,
    "type": "orcaleStake",
    "value": Object {
      "0x002002812b42601Ae5026344F0395E68527bb0F8": 234,
      "0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a": 123,
    },
  },
]
`;

exports[`actions subscribeAdmin 1`] = `
Array [
  Object {
    "id": 0,
    "type": "stakeSubscription",
  },
  Object {
    "id": 123,
    "type": "stakeSubscription",
  },
  Object {
    "index": 0,
    "type": "stakeStates",
    "value": Object {
      "canStake": false,
      "canUnstake": false,
      "lockMinTimeInSecond": 0,
      "maxStakePerOracle": 0,
      "minStakeAmt": 0,
      "totalAwardsToDistribute": 0,
    },
  },
  Object {
    "index": 0,
    "type": "stakeAddresses",
    "value": Object {
      "operator": "0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0",
      "priceFeedList": Array [
        "0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a",
        "0x002002812b42601Ae5026344F0395E68527bb0F8",
        "0x00476E55e02673B0E4D2B474071014D5a366Ed4E",
      ],
    },
  },
]
`;

exports[`actions subscribeAdmin 2`] = `30000`;

exports[`actions subscribeAdmin 3`] = `
Array [
  Object {
    "id": 0,
    "type": "stakeSubscription",
  },
  Object {
    "id": 123,
    "type": "stakeSubscription",
  },
  Object {
    "index": 0,
    "type": "stakeStates",
    "value": Object {
      "canStake": false,
      "canUnstake": false,
      "lockMinTimeInSecond": 0,
      "maxStakePerOracle": 0,
      "minStakeAmt": 0,
      "totalAwardsToDistribute": 0,
    },
  },
  Object {
    "index": 0,
    "type": "stakeAddresses",
    "value": Object {
      "operator": "0x415DE7Edfe2c9bBF8449e33Ff88c9be698483CC0",
      "priceFeedList": Array [
        "0x0022BFd6AFaD3408A1714fa8F9371ad5Ce8A0F1a",
        "0x002002812b42601Ae5026344F0395E68527bb0F8",
        "0x00476E55e02673B0E4D2B474071014D5a366Ed4E",
      ],
    },
  },
]
`;

exports[`actions subscriptionUpdate 1`] = `
Object {
  "id": 123,
  "type": "stakeSubscription",
}
`;

exports[`actions userAwardUpdate 1`] = `
Object {
  "index": 0,
  "type": "userAward",
  "value": Object {},
}
`;

exports[`actions userStakeUpdate 1`] = `
Object {
  "index": 0,
  "type": "userStake",
  "value": Object {},
}
`;
